### PR TITLE
initial working searchbox feature

### PR DESCRIPTION
**Pull request by rizary**
This PR follows what UI v2 behavior, however it still contains some performance issues such as:
1. When first typing, all packages generates first, then filtered into all packages which contain the searchbox value
2. That behavior also apply when we empty the searchbox, existing package list take several second to disappear.

### CODE REVIEW

**Comment by rizary** [comment]
The latest commit is based on @benkolera's revision on my first commit.

**Comment by benkolera** [comment]
@Rizary I think that you may not have pushed! :smile:

File: src-ui.v3/src/Main.hs [hunk 1]
```diff
@@ -136,6 +138,27 @@ bodyElement4 = do
       display (V.length <$> dynPackages0)
       text ")"
 
+    -- search box
+    searchInputE <- elAttr "div" ("class" =: "item search right clearfix") $ do
+              divClass "text" $ text "Package Search"
+              sVal0 <- inputElement $ def & inputElementConfig_elementConfig . elementConfig_initialAttributes .~ fold
+                ["class" =: "input-search","placeholder" =: "search..."]
+              debounce 0.1 $ _inputElement_input sVal0
```
Comment (benkolera) [inline src-ui.v3/src/Main.hs]: The UI felt a little bit better when I bumped this to 0.2 rather than 0.1, but see how it is for you because my perf testing was all over the shop. Haha

File: src-ui.v3/src/Main.hs [hunk 2]
```diff
@@ -136,6 +138,27 @@ bodyElement4 = do
       display (V.length <$> dynPackages0)
       text ")"
 
+    -- search box
```
Comment (benkolera) [inline src-ui.v3/src/Main.hs]: This section has some weird differences in indentation if we care. 4 spaces vs 2 spaces by the looks.

File: src-ui.v3/src/Main.hs [hunk 3]
```diff
@@ -137,28 +139,25 @@ bodyElement4 = do
...
-              constDyn $ elDynAttr "ul" (fmap (\t -> if t then "display" =: "none" else "display" =: "list-item") dynAttrBool) $ do
-                           _ <- listWithKey sDynMap $ \pId _ -> do -- m (Dynamic (Map k a))
-                                 el "li" $ elAttr "a" ("href" =: ("#/package/" <> pId)) $ text pId
-                           pure ()
-  
-      )
-    
+    let
+      textToJSString = JSS.pack . T.unpack
+      jsStringToText = T.pack . JSS.unpack
```
Comment (benkolera) [inline src-ui.v3/src/Main.hs]: We should kill the above to functions and use the conversions from Data.JSString.Text instead. I didn't know about those. That's my fault! :)

File: src-ui.v3/src/Main.hs (hunk 1 above)
Comment (rizary) [reply src-ui.v3/src/Main.hs]: I think we can comeback to this after we put style in the code.

File: src-ui.v3/src/Main.hs (hunk 2 above)
Comment (rizary) [reply src-ui.v3/src/Main.hs]: noted because at the top it is 4 spaces. I'll clean it up in the next commit.

**Comment by rizary** [comment]
Current bug:
1. if using `^` and `$` the bold effect does not show up ( I am sure that I already stripped the search box)
2. if you type package and then hold the backspace (in a long time) we will get an error:
```
Data.Text.breakOn: empty input
CallStack (from HasCallStack):
  error, called at ./Data/Text.hs:1857:18 in text-1.2.3.0-970Ii40pYwz7heH5hMz2Zy:Data.Text
```

and then the search box become malfunction.

File: src-ui.v3/src/Main.hs [hunk 4]
```diff
@@ -96,337 +98,381 @@ utc2unix x = ceiling (realToFrac (utcTimeToPOSIXSeconds x) :: Double)
...
+          
+  let
+    dynPackagesJss = V.toList . fmap (JSS.textToJSString . pkgNToText) <$> dynPackages0
+    calcMatches pkgs sJss =
+      if JSS.length sJss < 3
+      then []
+      else matches' pkgs sJss
+     where 
+      matches' p sJ
+        | JSS.isInfixOf "^" sJ = filter (JSS.isPrefixOf (JSS.dropWhile (=='^') sJ)) p
```
Comment (hvr) [inline src-ui.v3/src/Main.hs]:
```suggestion
        | Just sJ'  <- JSS.stripPrefix "^" sJ = filter (JSS.isPrefixOf sJ') p
```

File: src-ui.v3/src/Main.hs [hunk 5]
```diff
@@ -96,337 +98,381 @@ utc2unix x = ceiling (realToFrac (utcTimeToPOSIXSeconds x) :: Double)
...
+        | JSS.isInfixOf "$" sJ = filter (JSS.isSuffixOf (JSS.dropWhileEnd (=='$') sJ)) p
```
Comment (hvr) [inline src-ui.v3/src/Main.hs]:
```suggestion
        | Just sJ' <- JSS.stripSuffix "$" sJ = filter (JSS.isSuffixOf sJ') p
```

File: src-ui.v3/src/Main.hs [hunk 6]
```diff
@@ -96,337 +98,381 @@ utc2unix x = ceiling (realToFrac (utcTimeToPOSIXSeconds x) :: Double)
...
+        | otherwise            = filter (JSS.isInfixOf sJ) p
+    exactMatches pkgs' sJss = List.partition (==(JSS.dropWhileEnd (=='$') sJss)) pkgs'
+    matchesE = calcMatches <$> current dynPackagesJss <@> (JSS.textToJSString <$> searchInputE)
+  dynMatches <- holdDyn [] matchesE
+  dynSearch  <- holdDyn "" (JSS.textToJSString <$> searchInputE)
```
Comment (hvr) [inline src-ui.v3/src/Main.hs]: I suggest defining a 2nd `dynSearchFiltered` with any `$` or `^` removed; and use that one below in the code for the `<strong>`-splitting

File: src-ui.v3/src/Main.hs [hunk 7]
```diff
@@ -96,337 +98,381 @@ utc2unix x = ceiling (realToFrac (utcTimeToPOSIXSeconds x) :: Double)
...
+    matchesMapDyn = Map.fromList . fmap (\p -> (JSS.textFromJSString p,())) <$> (snd exactDyn)
+    exactMapDyn   = Map.fromList . fmap (\p -> (JSS.textFromJSString p,())) <$> (fst exactDyn)
+  --matchesMapDyn <- holdDyn Map.empty matchesMapE
+  _ <- el "ul" $ do
+        listWithKey exactMapDyn $ \eId _ ->
+          el "li" $ elAttr "a" ("href" =: ("#/package/" <> eId)) $ el "strong" $ text eId
+        listWithKey matchesMapDyn $ \pId _ ->
+          el "li" $ elAttr "a" ("href" =: ("#/package/" <> pId)) $ do
+            let 
+              tbFront txt txtS = T.breakOn txtS txt
```
Comment (hvr) [inline src-ui.v3/src/Main.hs]: would the code in this block be simpler if we define a (top-level) helper function

```hs
partitionOnString :: Text -> Text -> (Text,Text)
```

which would work like e.g.

```hs
>>> partitionOnString "bar" "foobardoo"
("foo","doo")

>>> partitionOnString "bar" "bar"
("","")
```


?

**Review by hvr — CHANGES_REQUESTED** [review]

File: src-ui.v3/cabal.project.local [hunk 8]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (benkolera) [inline src-ui.v3/cabal.project.local]: This file should not have been committed. Would be good to edit the gitignore to have this in there.

File: src-ui.v3/src/Main.hs [hunk 9]
```diff
@@ -731,11 +736,72 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
+
+searchBoxWidget :: forall t m. (SupportsServantReflex t m, MonadFix m, MonadIO m, MonadHold t m, PostBuild t m, DomBuilder t m, Adjustable t m, DomBuilderSpace m ~ GhcjsDomSpace)
+                => Dynamic t (Vector PkgN) 
+                -> m ()  
+searchBoxWidget dynPkgs0 = mdo
+  searchInputE <- elAttr "div" ("class" =: "item search right clearfix") $ do
+    divClass "text" $ text "Package Search"
+    sVal0 <- inputElement $ def & inputElementConfig_elementConfig . elementConfig_initialAttributes .~ fold ["class" =: "input-search","placeholder" =: "search..."]
+                                & inputElementConfig_setValue .~ clickPkgE
+    debounce 0.1 $ leftmost [clickPkgE, (_inputElement_input sVal0)]
```
Comment (benkolera) [inline src-ui.v3/src/Main.hs]: Does changing this to 0.2 or 0.3 make things smoother for you?

File: src-ui.v3/src/Main.hs [hunk 10]
```diff
@@ -731,11 +736,72 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
+    exactE <- listViewWithKey (matchesExact <$> mDyn) $ \eId _ -> do
+                (e, _) <- element "li" def $ elAttr "a" ("href" =: ("#/package/" <> eId)) $ el "strong" $ text eId
+                pure $ domEvent Click e
+    otherE <- listViewWithKey (matchesInfix <$> mDyn) $ \pId txt -> do
+                (e, _) <- element "li" def $ elAttr "a" ("href" =: ("#/package/" <> pId)) $ do
+                            dynText . fmap (^. _1) $ txt
+                            el "strong" $ dynText . fmap (^. _2) $ txt
+                            dynText . fmap (^. _3) $ txt
+                pure $ domEvent Click e
+    pure $ "" <$ leftmost [exactE, otherE]
```
Comment (benkolera) [inline src-ui.v3/src/Main.hs:818]: This is about the point where I'd just make this an EventWriter to tidy it up. It's borderline though. :slightly_smiling_face:

File: src-ui.v3/src/Main.hs [hunk 11]
```diff
@@ -731,11 +736,72 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
+  | otherwise            = filter (JSS.isInfixOf (stripSearch sJ))pkgs
+
+splitInfixPkg :: JSS.JSString -> JSS.JSString -> (T.Text, T.Text, T.Text)
+splitInfixPkg stripSJ pkg = (frontT, midT, backT)
+  where
+    textS               = JSS.textFromJSString stripSJ
+    (frontT, reminderT) = T.breakOn textS (JSS.textFromJSString pkg)
+    (midT, backT)       = T.breakOnEnd textS reminderT
+
+calcMatchesNew :: [JSS.JSString] -> JSS.JSString -> Matches
```
Comment (benkolera) [inline src-ui.v3/src/Main.hs]: This is looking a lot better now that it's all pure now!

File: src-ui.v3/src/Main.hs [hunk 12]
```diff
@@ -731,11 +736,72 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
+calcMatchesNew pkgs sJss = 
+  if JSS.length sJss < 3 
+  then matchesEmpty
+  else Matches { matchesInput = textS, matchesExact = exactMap, matchesInfix = othersMap}
+  where
+    textS                = JSS.textFromJSString sJss
+    stripSJ              = stripSearch sJss
+    (exactPkg, infixPkg) = List.partition (== stripSJ) pkgs
+    exactMap             = (Map.fromList . fmap (\p -> (JSS.textFromJSString p,()))) exactPkg
+    othersMap            = (Map.fromList . fmap (\p -> (JSS.textFromJSString p, splitInfixPkg stripSJ p))) (filterBySearchBox infixPkg sJss)
```
Comment (benkolera) [inline src-ui.v3/src/Main.hs]: This feels like it is probably slower than before because you do two passovers (one with the partition and one with the othersMap) the whole package set. 

If you make another function call calcMatch :: Text -> Text -> (Map Text (),Map Text (Text,Text,Text))

you could make exactMap and othersMap with a single pass like:
```
(exactMap,othersMap) <- foldMap (calcMatch textS) pkgs
```

That may be nicer and faster.

File: src-ui.v3/src/Main.hs (hunk 9 above)
Comment (rizary) [reply src-ui.v3/src/Main.hs]: 0.1 is way smoother when we have long text such as "facebook", I think i'm going to stick with 0.1

File: src-ui.v3/src/Main.hs (hunk 12 above)
Comment (rizary) [reply src-ui.v3/src/Main.hs]: it is faster, indeed. I have put it in my new commit. Thanks for your input!!

File: src-ui.v3/src/Main.hs [hunk 13]
```diff
@@ -777,11 +736,72 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
 
+stripSearch :: JSS.JSString -> JSS.JSString
+stripSearch sJ
+  | Just sJ'  <- JSS.stripPrefix "^" sJ = sJ'
+  | Just sJ' <- JSS.stripSuffix "$" sJ  = sJ'
+  | otherwise                           = sJ
+
+filterBySearchBox :: [JSS.JSString] -> JSS.JSString -> [JSS.JSString]
+filterBySearchBox pkgs sJ
+  | JSS.isInfixOf "^" sJ = filter (JSS.isPrefixOf (stripSearch sJ)) pkgs
```
Comment (hvr) [inline src-ui.v3/src/Main.hs]: checking `isInfixOf` and then stripping the prefix or suffix via `stripPrefix`/`stripSuffix` has two problems:

1. it mixes different semantics; isInfixOf will match in more cases than `stripPrefixOf`/`stripSuffixOf` will match on
2. it wastes on the benefit of using `stripPrefixOf`/`stripSuffixOf` in the first place: doing the check and the stripping in the very same step in an optimal way

in other words, the line above is more idiomatic/optimal written in the way below:

```suggestion
  | Just sJ' <- JSS.stripPrefixOf "^" sJ = filter (JSS.isPrefixOf sJ') pkgs
```

Comment (rizary) [reply src-ui.v3/src/Main.hs]: @hvr I think this file in this commit is already outdated. I have pushed my latest commit and have change `filterBySearchBox` entirely

File: src-ui.v3/src/Main.hs [hunk 14]
```diff
@@ -731,11 +737,83 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
+                                           then (Map.empty, Map.singleton (JSS.textFromJSString pkg) (splitInfixPkg sJ' pkg))
+                                           else (Map.empty, Map.empty)
+  | Just sJ' <- JSS.stripSuffix "$" sJss = if JSS.isSuffixOf sJ' pkg 
+                                           then (Map.empty, Map.singleton (JSS.textFromJSString pkg) (splitInfixPkg sJ' pkg))
+                                           else (Map.empty, Map.empty)
+  | otherwise                            = if JSS.isInfixOf sJss pkg
+                                           then (Map.empty, Map.singleton (JSS.textFromJSString pkg) (splitInfixPkg sJss pkg))
+                                           else (Map.empty, Map.empty)
+
+calcMatches :: [JSS.JSString] -> JSS.JSString -> Matches
```
Comment (hvr) [inline src-ui.v3/src/Main.hs:780]: It's more idiomatic to use guards here, e.g.

```hs
calcMatches :: [JSS.JSString] -> JSS.JSString -> Matches
calcMatches pkgs sJss 
  | JSS.length sJss < 3  = matchesEmpty
  | otherwise = Matches { matchesInput = textS, matchesExact = exactMap, matchesInfix = othersMap}
  where
    textS                = JSS.textFromJSString sJss
    (exactMap,othersMap) = F.foldMap (calcMatch sJss) pkgs
```

File: src-ui.v3/src/Main.hs [hunk 15]
```diff
@@ -95,338 +99,340 @@ utc2unix :: UTCTime -> Int
...
+          deleteTag0 :: Event t (Map.Map T.Text T.Text) <- listViewWithKey result $ \tId _ -> do
+            el "li" $ do
+              el "span" $ text tId
+              delEv <- rmTagButton_ tId pn
+              pure $ tagNToText <$> delEv
+
+          addTag0 <- elClass "form" "form" $ do
+            el "p" $ text "Tag : "
+            tagName <- textInput iCfg
+            tagButton <- clickElement_ "button" "Add Tag"
```
Comment (hvr) [inline src-ui.v3/src/Main.hs:391]: why this seemingly unrelated change in this PR?

File: src-ui.v3/src/Main.hs [hunk 16]
```diff
@@ -731,11 +737,83 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
+  | otherwise                           = sJ
 
+splitInfixPkg :: JSS.JSString -> JSS.JSString -> (T.Text, T.Text, T.Text)
+splitInfixPkg stripSJ pkg = (frontT, midT, backT)
+  where
+    textS               = JSS.textFromJSString stripSJ
+    (frontT, reminderT) = T.breakOn textS (JSS.textFromJSString pkg)
+    (midT, backT)       = T.breakOnEnd textS reminderT
+
+calcMatch :: JSS.JSString  -> JSS.JSString -> (Map.Map T.Text (), Map.Map T.Text (T.Text,T.Text,T.Text))
```
Comment (hvr) [inline src-ui.v3/src/Main.hs]: it's more idiomatic to write

```hs
calcMatch :: JSS.JSString  -> JSS.JSString -> (Map.Map T.Text (), Map.Map T.Text (T.Text,T.Text,T.Text))
calcMatch sJss pkg
  | stripSJ == pkg  = (Map.singleton (JSS.textFromJSString pkg) (), Map.empty)
  | otherwise  = filterPkgSearch stripSJ pkg
  where
    stripSJ = stripSearch sJss
```

File: src-ui.v3/src/Main.hs [hunk 17]
```diff
@@ -731,11 +737,83 @@ pkgTagList m = Map.fromListWith (List.++) $ do
...
+  in 
+    case stripSJ == pkg of
+      True  -> (Map.singleton (JSS.textFromJSString pkg) (), Map.empty)
+      False -> filterPkgSearch stripSJ pkg
+
+filterPkgSearch :: JSS.JSString -> JSS.JSString -> (Map.Map T.Text (), Map.Map T.Text (T.Text,T.Text,T.Text))
+filterPkgSearch sJss pkg
+  | Just sJ' <- JSS.stripPrefix "^" sJss = if JSS.isPrefixOf sJ' pkg 
+                                           then (Map.empty, Map.singleton (JSS.textFromJSString pkg) (splitInfixPkg sJ' pkg))
+                                           else (Map.empty, Map.empty)
```
Comment (hvr) [inline src-ui.v3/src/Main.hs]: btw, you can also just write

```suggestion
                                           else mempty
```

which is the same as `(mempty, mempty)` and thus also `(Map.empty, Map.empty)`

File: src-ui.v3/src/PkgId.hs [hunk 18]
```diff
@@ -144,4 +147,16 @@ verToText (Ver x) = T.pack . Ver.showVersion . Ver.makeVersion $ x
...
+
+data Matches = Matches 
+  { matchesInput :: Text
+  , matchesExact :: Map.Map Text ()
+  , matchesInfix :: Map.Map Text (Text,Text,Text)
+  }
+  deriving (Eq,Ord)
+
+matchesEmpty :: Matches
+matchesEmpty = Matches { matchesInput = T.empty :: T.Text, matchesExact = Map.empty, matchesInfix = Map.empty}
```
Comment (hvr) [inline src-ui.v3/src/PkgId.hs]: why the redundant `:: T.Text` type-sig?

**Review by hvr — CHANGES_REQUESTED** [review]
generally, you can drop all the `T.` module prefixes from the `T.Text` occurences in `Main.hs`

File: src-ui.v3/src/PkgId.hs (hunk 18 above)
Comment (rizary) [reply src-ui.v3/src/PkgId.hs]: i'll remove it. At first it is `""` then I found the `T.empty` but forgot to remove it.

File: src-ui.v3/src/Main.hs (hunk 15 above)
Comment (rizary) [reply src-ui.v3/src/Main.hs:391]: I just update the `button_` to make it more general use.

File: src-ui.v3/src/Main.hs [hunk 19]
```diff
@@ -793,7 +793,7 @@ searchBoxWidget dynPkgs0 = mdo
     divClass "text" $ text "Package Search"
     sVal0 <- inputElement $ def & inputElementConfig_elementConfig . elementConfig_initialAttributes .~ fold ["class" =: "input-search","placeholder" =: "search..."]
                                 & inputElementConfig_setValue .~ clickPkgE
-    debounce 0.1 $ leftmost [clickPkgE, (_inputElement_input sVal0)]
+    debounce 0.3 $ leftmost [clickPkgE, (_inputElement_input sVal0)]
```
Comment (rizary) [inline src-ui.v3/src/Main.hs:796]: @benkolera I think 0.3 is better after several attempts on search. Thanks, Ben.

**Comment by hvr** [comment]
We'll deal with cleaning up the code in a follow-up PR
